### PR TITLE
fix(compiler-cli): shorten resolved module name in fileNameToModuleName to npm package name for typings

### PR DESCRIPTION
**Pull request by gregmagolan**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently fileNameToModuleName resolves typings files to module names as such: `.../node_modules/@angular/core/core.d.ts` => `@angular/core/core`, `.../node_modules/rxjs/index.d.ts` => `rxjs/index`.

Issue Number: N/A

## What is the new behavior?
If the typings in the npm package matches the imported file then the module name is resolved to the npm package name as such: `.../node_modules/@angular/core/core.d.ts` => `@angular/core`, `.../node_modules/rxjs/index.d.ts` => `rxjs`.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: integration/bazel/BUILD.bazel [hunk 1]
```diff
@@ -2,15 +2,21 @@ package(default_visibility = ["//visibility:public"])
...
-        ["node_modules/**/*"],
-        # Exclude directories that commonly contain filenames which are
-        # illegal bazel labels
-        exclude = [
-            # e.g. node_modules/adm-zip/test/assets/attributes_test/New folder/hidden.txt
-            "node_modules/**/test/**",
-            # e.g. node_modules/xpath/docs/function resolvers.md
-            "node_modules/**/docs/**",
-        ],
+    srcs = glob(["/".join([
```
Comment (alexeagle) [inline integration/bazel/BUILD.bazel]: let's keep it the other way (will be needed for hermeticity) and add `node_modules/rxjs/**` to the excludes with a comment explaining why we need to test for this case

File: packages/compiler-cli/src/transformers/compiler_host.ts [hunk 2]
```diff
@@ -211,8 +212,20 @@ export class TsCompilerAotCompilerTypeCheckHostAdapter implements ts.CompilerHos
...
       moduleName = dotRelative(path.dirname(containingFile), importedFile);
-    } else if (importedFilePackagName) {
+    } else if (importedFilePackageName) {
       moduleName = stripNodeModulesPrefix(importedFile);
+      if (origImportedFile.endsWith('.d.ts')) {
+        // the moduleName for these typings could be shortented to the npm package name
+        // if the npm package typings matches the importedFile
+        try {
+          const modulePath = importedFile.substring(0, importedFile.length - moduleName.length) + importedFilePackageName;
+          const packageJson = require(modulePath + '/package.json');
```
Comment (alexeagle) [inline packages/compiler-cli/src/transformers/compiler_host.ts]: this probably gets called a lot, I think it's probably worth memoizing this lookup for performance

**Review by alexeagle — APPROVED** [review]

File: packages/compiler-cli/src/transformers/compiler_host.ts [hunk 3]
```diff
@@ -211,12 +217,28 @@ export class TsCompilerAotCompilerTypeCheckHostAdapter implements ts.CompilerHos
...
+        // if the npm package typings matches the importedFile
+        try {
+          const modulePath = importedFile.substring(0, importedFile.length - moduleName.length) +
+              importedFilePackageName;
+          const packageJson = require(modulePath + '/package.json');
+          const packageTypings = path.posix.join(modulePath, packageJson.typings);
+          if (packageTypings === originalImportedFile) {
+            moduleName = importedFilePackageName;
+          }
+        } catch (e) {
```
Comment (alxhub) [inline packages/compiler-cli/src/transformers/compiler_host.ts]: Leave a comment explaining why ignoring the error is okay.

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/src/transformers/compiler_host.ts (hunk 3 above)
Comment (gregmagolan) [reply packages/compiler-cli/src/transformers/compiler_host.ts]: Done

**Pull request closed without merging**